### PR TITLE
Backend Membership Renewal using PriceSets

### PR DESCRIPTION
**Pull request by mlutfy**
Overview
----------------------------------------

This is another attempt to #11182 and #9216 to add support for renewing memberships using a PriceSet on the backend.

Test cases:

- Backend New Membership with PriceSet
  - Renew it using the same PriceSet options
  - Renew it using a different membership type from the same PriceSet
- Backend New Membership without Priceset
  - Renew it without a PriceSet
  - Renew it with a PriceSet (this has known issues)
- Backend New Membership with Credit Card and with Price Set
- Backend Renew Membership with Credit Card and with Price Set
- etc

It addresses some issues raised in #11182, such as:

- Changing the membership type during renewal now updates the membership record (ex: from "student" to "general")
- When renewing, the default price set selections (from last time) are set by default.

Before
----------------------------------------

A membership created with a PriceSet cannot be renewed from the backend (by an admin). They have to use workarounds to renew using a ContributionPage (which then sets the payment as pending, has to be manually set to completed).

After
----------------------------------------

Backend renewal possible.

Comments
----------------------------------------

Known issues:

- Sometimes, when renewing offline without credit card and priceset, and changing the selections
  - the "record contribution" checkbox is sometimes not set (and must be set, according some obscure error message, I did not investigate why)
  - the total amount is sometimes not set correctly, and we have to trigger a change in the priceset so that it gets updated

(I'm not really sure because I did so many tests and I'm not sure I can reproduce then now)

Todo: add tests, which were part of the original PR?

cc @masetto @nganivet @monishdeb (apologies for the blast from the past, but fyi if still interested)

### CODE REVIEW

File: CRM/Member/Form.php [hunk 1]
```diff
@@ -365,6 +380,220 @@ public function buildQuickForm() {
...
+   *   (ref.) an assoc array of name/value pairs.
+   *
+   * @param array $files
+   * @param object $self
+   *
+   * @throws CiviCRM_API3_Exception
+   * @return bool|array
+   *   mixed true or array of errors
+   */
+  public static function formRule($params, $files, $self) {
```
Comment (mlutfy) [inline CRM/Member/Form.php:396]: Moved as-is from CRM_Member_Form_Membership

File: CRM/Member/Form.php [hunk 2]
```diff
@@ -657,4 +886,38 @@ protected function getCardTypeID(): ?int {
     return CRM_Core_PseudoConstant::getKey('CRM_Core_BAO_FinancialTrxn', 'card_type_id', $this->getSubmittedValue('credit_card_type'));
   }
 
+  /**
+   * Get parameters for membership create for all memberships to be created.
+   *
+   * @return array
+   * @throws \CRM_Core_Exception
+   */
+  protected function getMembershipParameters(): array {
```
Comment (mlutfy) [inline CRM/Member/Form.php:895]: Moved as-is from CRM_Member_Form_Membership

File: CRM/Member/Form/Membership.php [hunk 3]
```diff
@@ -154,11 +154,6 @@ public function preProcess() {
     $this->assign('formClass', 'membership');
     parent::preProcess();
 
-    // get price set id.
-    $this->_priceSetId = $_GET['priceSetId'] ?? NULL;
-    $this->set('priceSetId', $this->_priceSetId);
-    $this->assign('priceSetId', $this->_priceSetId);
-
```
Comment (mlutfy) [inline CRM/Member/Form/Membership.php:161]: Now set in the parent

File: CRM/Price/BAO/PriceSet.php [hunk 4]
```diff
@@ -1633,14 +1633,14 @@ protected static function addPriceFieldsToForm(CRM_Core_Form $form, $feeBlock, b
         !$validFieldsOnly
       ) {
         $options = $field['options'] ?? NULL;
-        if ($className == 'CRM_Contribute_Form_Contribution_Main' && $component = 'membership') {
+        if ($className == 'CRM_Contribute_Form_Contribution_Main' && $component == 'membership') {
```
Comment (mlutfy) [inline CRM/Price/BAO/PriceSet.php:1636]: Fixing this was not necessary, and maybe should be in a separate PR. Looks like a bug?

**Comment by mlutfy** [comment]
Test errors:

```

Installing build0test_abzw7 database
not ok 2308 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmit
not ok 2309 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitWithTax
not ok 2310 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitWithTaxOfZero
not ok 2311 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitChangeType
not ok 2312 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitRecur
not ok 2313 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitRecurCompleteInstant
not ok 2314 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitRecurCompleteInstantWithMail with data set #0 ('.')
not ok 2315 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitRecurCompleteInstantWithMail with data set #1 (',')
not ok 2316 - Error: CRM_Member_Form_MembershipRenewalTest::testMailOutput
not ok 2317 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitPayLater
not ok 2318 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitPayLaterWithBilling
not ok 2319 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitComplete
not ok 2320 - Error: CRM_Member_Form_MembershipRenewalTest::testSubmitRenewExpired

Installing build0test_abzw7 database
not ok 2321 - Error: CRM_Member_Form_MembershipTest::testFormRuleEmptyContact
not ok 2322 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingEarlyStart
not ok 2323 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingEarlyEnd
not ok 2324 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingEndNoStart
not ok 2325 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingLifetimeEnd
not ok 2326 - Error: CRM_Member_Form_MembershipTest::testFormRulePermanentOverrideWithNoStatus
not ok 2327 - Error: CRM_Member_Form_MembershipTest::testFormRuleUntilDateOverrideWithValidOverrideEndDate
not ok 2328 - Error: CRM_Member_Form_MembershipTest::testFormRuleUntilDateOverrideWithNoOverrideEndDate
not ok 2329 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingJoin1MonthFromNow
not ok 2330 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingJoinToday
not ok 2331 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingJoin1MonthAgo
not ok 2332 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingJoin6MonthsAgo
not ok 2333 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingJoin1YearAgo
not ok 2334 - Error: CRM_Member_Form_MembershipTest::testFormRuleRollingJoin2YearsAgo
not ok 2335 - Error: CRM_Member_Form_MembershipTest::testFormRuleFixedJoin6MonthsAgo
```

File: templates/CRM/Member/Form/MembershipRenewal.tpl [hunk 5]
```diff
@@ -8,6 +8,9 @@
  +--------------------------------------------------------------------+
 *}
 {* this template is used for renewing memberships for a contact  *}
+{if $priceSetOnly}
```
Comment (eileenmcnaughton) [inline templates/CRM/Member/Form/MembershipRenewal.tpl:11]: If you are doing this to load a new set of values onto the form (ie an overload) we are working to put them in separate form / template classes with their own paths rather than having templates that are basically big IF- elses

Comment (mlutfy) [reply templates/CRM/Member/Form/MembershipRenewal.tpl:11]: @eileenmcnaughton Any examples of forms that did that recently? (I'd also like to switch this form to `snippet=json`, because `snippet=4` is a pain for extensions who might add custom JS/vars)

File: CRM/Price/BAO/PriceSet.php [hunk 6]
```diff
@@ -1690,4 +1691,41 @@ public static function getExtendsOptions(): array {
...
 
+  /**
+   * Get priceset ID of selected entity ID
+   *
+   * @param int $entityID
+   * @param string $entityTable
+   *
+   * @return int Price set ID
+   */
+  public static function getPriceSetFromEntityID($entityID, $entityTable = 'civicrm_contribution') {
```
Comment (eileenmcnaughton) [inline CRM/Price/BAO/PriceSet.php:1702]: This shouldn't be needed - you already have a BAO_Order object so getPriceSetID() should work on it...

**Comment by eileenmcnaughton** [comment]
I suppose at a high level I never quite got to the bottom of what the mechanics should be - the goal is obviously the renewal creates an order & then adds a payment to complete it. When I last dug in I couldn't quite get my head around some of what was happening in this form but I think it's calling the BAO membership create in a scary magic way that I didn't fully understand & which isn't necessarily consistent with other forms.

 I've spent so much time trying to unravel this stuff adding more complexity scares me. The goal is to make this possible via form builder - but that needs api/ action clarity

**Comment by mlutfy** [comment]
I understand the feeling. I was kind of hoping to get to a point where the actual code added/duplicated was somewhat minor, but the tests aren't happy about it either.

I'm not in a rush to move this forward - the client didn't want to fund it, I was mostly curious, but I appreciate the review.

I might poke around at some point to simplify bits or propose a separate to move the functions, so that we could run this as a small patch that will be easier to maintain. We get a lot of requests for this, but we usually find workarounds.

**Comment by eileenmcnaughton** [comment]
@mlutfy yeah - the underlying issue is there is some nasty complex stuff going on in there & when you open the lid it all comes out - oh hang on was that pandora's box I was thinking of. But it would be good to do another round of trying to simplify. I've been focussed on event records lately - trying to update the workflow templates & get the flow to be notice free but had membership on my agenda after that - albeit my focus was gonna be online
